Add documentation for experimental features. (#10865)
matrix-org/synapse: docs/development/experimental_features.md

# Implementing experimental features in Synapse

It can be desirable to implement "experimental" features which are disabled by
default and must be explicitly enabled via the Synapse configuration. This is
applicable for features which:

> * Are unstable in the Matrix spec (e.g. those defined by an MSC).

Reviewer: ```suggestion
* Are unstable in the Matrix spec (e.g. those defined by an MSC that has not yet been merged).
```

> * Developers are not confident are ready for use by general Synapse administrators/users

Reviewer: ```suggestion
* Developers are not confident in their use by general Synapse administrators/users
```

Possibly?

  (e.g. a feature is incomplete, buggy, performs poorly, or needs further testing).

> Note that this does *not* include features which are not expected to be desirable

Reviewer: ```suggestion
Note that this only really applies to features which are expected to be desirable
```

An attempt at eliminating double negatives.

to a broad audience. The [module infrastructure](../modules/index.md) should be

> investigated for non-standard features.

Reviewer: Is the suggestion here that some module features should be hidden behind experimental config flags, or?

Author: The suggestion is that some features should be built as modules, not as things behind experimental config flags.

Author: Maybe it is clearer to just remove this? Not sure, was trying to give people other options though.

Reviewer: Aha OK! In that case, I might just slightly reword this to:

```suggestion
to a broad audience. The [module infrastructure](../modules/index.md) should
instead be investigated for non-standard features.
```

Guarding experimental features behind configuration flags should help with some
of the following scenarios:

* Ensure that clients do not assume that unstable features exist (failing
  gracefully if they do not).
* Unstable features do not become de-facto standards and can be removed
  aggressively (since only those who have opted-in will be affected).
* Ease finding the implementation of unstable features in Synapse (for future
  removal or stabilization).
* Ease testing a feature (or removal of feature) due to enabling/disabling without
  code changes. It also becomes possible to ask for wider testing, if desired.

Experimental configuration flags should be disabled by default (requiring Synapse
administrators to explicitly opt-in), although there are situations where it makes
sense (from a product point-of-view) to enable features by default. This is
expected and not an issue.

> It is not a requirement that experimental features require a configuration flag,
> but if one should be included if unsure.

Reviewer: ```suggestion
It is not a requirement for experimental features to be behind a configuration flag,
but one should be used if unsure.
```
